### PR TITLE
Issue #123: implement turntable mouse navigation

### PR DESCRIPTION
**Pull request by dynamodan**
I took a little time to implement this.  It wasn't very difficult.  This comes with config checkbox (disabled by default).

### CODE REVIEW

File: src/solvespace.cpp [hunk 1]
```diff
@@ -70,6 +70,8 @@ void SolveSpaceUI::Init() {
     drawBackFaces = CnfThawBool(true, "DrawBackFaces");
     // Check that contours are closed and not self-intersecting
     checkClosedContour = CnfThawBool(true, "CheckClosedContour");
+    // Use turntable mouse navigation
+    checkTurntableNav = CnfThawBool(false, "CheckTurntableNav");
```
Comment (whitequark) [inline src/solvespace.cpp]: "check" in "checkClosedContour" doesn't mean "checkbox", it's a part of the option name

File: src/mouse.cpp [hunk 2]
```diff
@@ -129,8 +129,14 @@ void GraphicsWindow::MouseMoved(double x, double y, bool leftDown,
 
         if(!(shiftDown || ctrlDown)) {
             double s = 0.3*(PI/180)*scale; // degrees per pixel
-            projRight = orig.projRight.RotatedAbout(orig.projUp, -s*dx);
-            projUp = orig.projUp.RotatedAbout(orig.projRight, s*dy);
+            if(SS.checkTurntableNav) {
+                projRight = orig.projRight.RotatedAbout(Vector::From(0, 1, 0), -s*dx);
+                projUp = orig.projUp.RotatedAbout(Vector::From(orig.projRight.x, orig.projRight.y, 0), s*dy); // lock the Z to vertical
+			} else {
```
Comment (whitequark) [inline src/mouse.cpp]: please use [our coding style](https://github.com/solvespace/solvespace/blob/master/CONTRIBUTING.md#coding-style)

**Comment by dynamodan** [comment]
Ok so commit d11ba149d31994abd560 addresses the variable naming and the coding style (I think).  I think it was the one line being more than 100 characters long, but you didn't specify.  Not sure if you will like the coding style yet. :)

**Comment by whitequark** [comment]
@dynamodan There are more tabs in that commit...

Anyway, other than style issues I like it. Unfortunately we still have to sort out a CLA for external contributions so I can't merge it just yet, but I'm on it.

**Comment by whitequark** [comment]
Hm, don't you think the Y axis is not the right one? The initial sketch is on the XY plane, so if you're extruding it then the 'turntable' will not rotate parallel the extrusion vector.

**Comment by dynamodan** [comment]
The build failed due to some write permission in the mac part of the travis CI:
`[ 72%] Built target solvespace
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f src/CMakeFiles/solvespace-dmg.dir/build.macd /Users/travis/build/solvespamake[2]: write error
make[1]: *** [src/CMakeFiles/solvespace-dmg.dir/all] Error 1
make[1]: write error
make: *** [all] Error 1
+echo 'I can'\''t believe how deep the Travis brokenness goes. Retrying...'
./.travis/build-macos.sh: line 13: echo: write error: Resource temporarily un
The command "if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then ./.travis/build-macos.sh; fi" exited with 1.
Done. Your build exited with 1.
/Users/travis/build.sh: line 148: shell_session_update: command not found`

Doesn't have anything to do with this pull.

**Comment by dynamodan** [comment]
`Hm, don't you think the Y axis is not the right one?`
I thought about that... One thing I realized is that when I export to triangle mesh for Blender, then the part is lying on its side.  I may try to get this to match Blender's coordinate system where the Z axis is "up".  Is that what you were referring to?  I also noticed that by default, Y seems to be "up" in solvespace.  Or maybe it's just that the default view is from the top, which makes it seem so.

**Comment by Evil-Spirit** [comment]
@dynamodan, just use the z axis for upvectore as all the normal guys.

**Comment by dynamodan** [comment]
@Evil-Spirit I tried to make Z (the blue axis) be vertical, but it causes some strange bugs.  By default, Y is vertical, I'm sure of it, because of the way the horizontal and vertical constraints are messed up in some work planes when Z is "up".  So... Y is up for now.

**Comment by whitequark** [comment]
@dynamodan Horizontal and vertical constraints have nothing to do with XYZ (i.e. the global coordinate system). They are defined in reference to the workplane they're in.

**Comment by Evil-Spirit** [comment]
@whitequark, I can try to fix this for Z axis

**Comment by dynamodan** [comment]
Here's an example of what I was referring to with the H and V constraints.  The sketching is being done in the XY plane, with the Z axis pointing up.  The constraints work perfectly well -- the lines stay constrained.  But look how the H's are on the vertical lines, and the V's are on the horizontal lines.  It's not really "messed up" I guess.
solvespace - new sketch _003

**Comment by dynamodan** [comment]
Here's a screenshot of what I mean by the constraints being messed up.  They aren't really -- the constraints still work perfectly to constrain the lines straight.  The sketching is being done in the ZX workplane, with Z pointing "up", that is, toward the top of the screen.
solvespace - new sketch _004

**Comment by whitequark** [comment]
Yeah, that only affects the text orientation.

**Comment by dynamodan** [comment]
Well, not just the H and V lying on their sides, but the H is on the vertical segment, and the V is on the horizontal segment.  It really confused me the other day, LOL!  Anyway, in the context of turntable mouse navigation, I'm happy.

**Comment by whitequark** [comment]
@dynamodan Actually, now that I look closer, there's no issue. Try pressing F2 (without any turntable mode enabled). You'll see that horizontal will become what you expect from the constraints.

**Comment by dynamodan** [comment]
I've discovered a new bug in my turntable nav implementation.  When the view is aligned to a workplane that tilts the locked axis horizontally, the view becomes locked to middlebutton+drag mouse navigation.  (Other movements still work, such as ctrl+middlebutton+drag and shift+middlebutton+drag.)

So in regard to this pull request, don't pull and merge just yet.

**Comment by dynamodan** [comment]
This is done now -- I've fixed the issue of Z not staying vertical when some workplanes are active.  Also, over the last few weeks I've been periodically pulling and merging from the master branch and testing to keep my fork fresh so whenever you're ready it should be good to pull.

**Comment by whitequark** [comment]
@dynamodan OK. I'll take a look soon. We unfortunately still don't have a CLA so I can't merge yet.

**Comment by whitequark** [comment]
We have a CLA now so hopefully this will be merged soon!

**Comment by dynamodan** [comment]
That's good news!  Did you do the merge yet?  I just tonight merged from the latest (16540b1b2c at this moment) commit and successfully built in my fork.  I try to do this periodically.

**Comment by whitequark** [comment]
Still waiting on CLA stuff...

**Comment by whitequark** [comment]
Thank you for your contribution, and welcome. The SolveSpace project has a Contributor License Agreement; in order for me to be able to merge your pull request, please sign it at https://cla-assistant.io/solvespace/solvespace. It will take only a small amount of your time.

**Comment by dynamodan** [comment]
I signed the CLA.  I'll pull the latest and get my branch up to date, seems like that will make it easier to pull the turntable stuff.

File: src/confscreen.cpp [hunk 3]
```diff
@@ -84,6 +84,23 @@ void TextWindow::ScreenChangeBackFaces(int link, uint32_t v) {
...
+        SS.GW.AnimateOnto(
+            Quaternion::From(
+                Vector::From(-1,0,0),
+                Vector::From(0,0,1)
+            ),
+            SS.GW.offset
+        );
+    }
+    
+    InvalidateGraphics(); // not sure if this is needed after the AnimateOnto, but hey
```
Comment (whitequark) [inline src/confscreen.cpp]: Nope, AnimateOnto implies InvalidateGraphics.

**Comment by dynamodan** [comment]
I'll try to see what changed in extlib/libdxfrw, but they weren't my changes that I know of.  I believe I had simply run
```
git submodule update --init extlib/libdxfrw
```
as directed in the build instructions.  Something could have changed in that submodule?

**Comment by dynamodan** [comment]
I excluded extlib/libdxfrw.  Not sure if I did it correctly, as now the appveyor and travis builds fail.

**Comment by whitequark** [comment]
No, you just deleted the submodule...

**Comment by dynamodan** [comment]
I guess I don't understand how to do this.  When I checkout your branch, the extlib/libdxfrw folder is empty.  When I checkout my branch, the extlib/libdxfrw folder is empty.  Same, same!  But when I do `git submodule update --init extlib/libdxfrw` in my branch, I get `error: pathspec 'extlib/libdxfrw' did not match any file(s) known to git`

**Comment by whitequark** [comment]
Thanks! I'll take a closer look soon and merge--there isn't obviously wrong that I can see.

**Comment by whitequark** [comment]
Merged in master, thank you for your contribution and patience.

**Pull request closed without merging**